### PR TITLE
feat: introduce service to manage state

### PR DESCRIPTION
**Pull request by heikkipesonen**
add service to keep track of the current state of the user-centrics as react can be initialized long after the events have been fired

also increase timeout since 5s is a bit too short on dev.

### CODE REVIEW

File: src/components/UsercentricsProvider.tsx [hunk 1]
```diff
@@ -28,27 +25,19 @@ interface UsercentricsProviderProps {
...
- * () => <UsercentricsProvider windowEventName="ucEvent"><App /></UsercentricsProvider>
- *
- * @example <caption>Custom timeout in milliseconds</caption>
- * () => <UsercentricsProvider timeout={100} windowEventName="ucEvent"><App /></UsercentricsProvider>
- */
 export const UsercentricsProvider: FC<UsercentricsProviderProps> = ({
     children,
     strictMode = false,
-    timeout = 5000,
+    timeout = 30000,
```
Comment (iiroj) [inline src/components/UsercentricsProvider.tsx:31]: In my opinion 30 seconds is too long in production. Since this can be adjusted via the `prop` for development purposes, is this a necessary change?

```tsx
<UsercentricsProvider timeout={isProduction ? 5000 : 30000}>
```

Comment (heikkipesonen) [reply src/components/UsercentricsProvider.tsx:31]: i think the same happens occasionally in prod, the loading can take a while...

Comment (heikkipesonen) [reply src/components/UsercentricsProvider.tsx:31]: maybe we reduce this to 15s?

Comment (iiroj) [reply src/components/UsercentricsProvider.tsx]: Can we just remove this change from the PR? It's technically unrelated to the main feature, and the value can already be overridden via the prop.

File: src/components/UsercentricsProvider.tsx [hunk 2]
```diff
@@ -28,27 +25,19 @@ interface UsercentricsProviderProps {
...
- */
 export const UsercentricsProvider: FC<UsercentricsProviderProps> = ({
     children,
     strictMode = false,
-    timeout = 5000,
+    timeout = 15000,
     windowEventName,
 }) => {
+    const [state, setState] = useState<ServiceState>(UserCentericsService.state)
+    UserCentericsService.subscribe(setState)
```
Comment (iiroj) [inline src/components/UsercentricsProvider.tsx:35]: This calls `UserCentericsService.subscribe()` on every render of the provider. Should we move it to a `useEffect` instead? I'm pretty sure we can leave the deps array empty, so it will only run once.

File: src/services/user-centrics-service.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+if (typeof window !== 'undefined') {
+    window.addEventListener('UC_UI_CMP_EVENT', onEvent)
+    window.addEventListener('UC_UI_INITIALIZED', () => setIsInitialized(true))
+}
+
+export const UserCentericsService = {
+    state,
+    subscribe,
+}
```
Comment (iiroj) [inline src/services/user-centrics-service.ts:51]: We could encapsulate the `state` and `subscriptions` states by making this a factory. This would be a bit cleaner in my opinion. Something like:

```ts
const usercentrics = createUsercentricsService()
```

where all things currently on module-level are moved into the function body.

What do you think?

File: src/components/UsercentricsProvider.tsx (hunk 2 above)
Comment (iiroj) [reply src/components/UsercentricsProvider.tsx:35]: I wonder if it would be technically cleaner to also supply `UserCentericsService.unsubscribe(<some-id>)`... 🤔 Maybe the use-case doesn't really require it.

Comment (heikkipesonen) [reply src/components/UsercentricsProvider.tsx:35]: the use-effect is a good point, ill make it

File: src/components/UsercentricsProvider.tsx (hunk 1 above)
Comment (heikkipesonen) [reply src/components/UsercentricsProvider.tsx]: whatever, other projects will have to take this into account as well.

File: src/components/UsercentricsProvider.tsx [hunk 4]
```diff
@@ -2,12 +2,9 @@ import { FC, ReactNode, useMemo } from 'react'
 import React, { useCallback, useEffect, useState } from 'react'
 
 import { UsercentricsContext } from '../context.js'
-import type { UCUICMPEvent, UCWindow } from '../types.js'
-import { UCUICMPEventType } from '../types.js'
+import { ServiceState, UserCentericsService } from '../services/user-centrics-service.js'
```
Comment (iiroj) [inline src/components/UsercentricsProvider.tsx:5]: This is nitpick, but let's normalize the capitalization to `Usercentrics`, since that's how the company name is written.

File: src/services/user-centrics-service.ts (hunk 3 above)
Comment (iiroj) [reply src/services/user-centrics-service.ts:51]: It is currently a singleton yes, but does it have to be? I don't really have a strong opinion on this, so just a thought.

Comment (iiroj) [reply src/services/user-centrics-service.ts:51]: Let's keep it as a singleton for now.

Comment (iiroj) [inline src/services/user-centrics-service.ts:51]: Should we add some JSDoc comments for this?

Comment (heikkipesonen) [reply src/services/user-centrics-service.ts:56]: feel free to contribute if you see some benefit for those, ive been used to ts types 🤔

File: src/components/UsercentricsProvider.tsx (hunk 4 above)
Comment (heikkipesonen) [reply src/components/UsercentricsProvider.tsx]: yep, fixed.

File: src/services/user-centrics-service.ts (hunk 3 above)
Comment (iiroj) [reply src/services/user-centrics-service.ts]: Maybe we can go with something like:

```ts
/** State management singleton for internal use. This should not be used directly in applications. */
```

Comment (heikkipesonen) [reply src/services/user-centrics-service.ts]: added comment.

**Review by iiroj — APPROVED** [review]
I'm very happy with the PR, thanks! Let's merge this after some manual testing.

One last thing, we can squash&merge the two commits but could maybe reword the commit message a bit to make it more clear what the improvement is?

Maybe something like:

```
feat: detect Usercentrics events earlier by extracting singleton service
```
